watchtower/wtclient/client.go

```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
+// closed in the DB.
+//
+// NOTE: This method MUST be run as a goroutine.
+func (c *TowerClient) handleChannelCloses(chanSub subscribe.Subscription) {
+	defer c.wg.Done()
+
+	c.log.Tracef("Starting channel close handler")
+	defer c.log.Tracef("Stopping channel close handler")
+
+	for {
+		select {
+		case event, ok := <-chanSub.Updates():
+			if !ok {
+				log.Debugf("channel notifier has exited")
+				return
+			}
+
+			switch event := event.(type) {
+			case channelnotifier.ClosedChannelEvent:
+				closeInfo := event.CloseSummary
+				chanID := lnwire.NewChanIDFromOutPoint(
+					&closeInfo.ChanPoint,
+				)
+
+				// We only care about channels registered with
+				// the tower client.
+				c.backupMu.Lock()
+				_, ok := c.summaries[chanID]
+				c.backupMu.Unlock()
+				if !ok {
+					continue
+				}
+
+				sl, err := c.cfg.DB.MarkChannelClosed(
+					chanID, closeInfo.CloseHeight,
+				)
+				if err != nil {
+					log.Errorf("could not handle closed "+
+						"channel(%s): %v", chanID, err)
+					continue
+				}
+
+				for _, sID := range sl {
+					delay, err := newRandomDelay(
+						c.cfg.SessionCloseRange,
+					)
+					if err != nil {
+						log.Errorf("could not "+
+							"generate delay: %v",
+							err)
+					}
+
+					deleteHeight := closeInfo.CloseHeight +
+						delay
+
+					item := &sessionCloseItem{
+						sessionID:    sID,
+						deleteHeight: deleteHeight,
+					}
+
+					c.closableSessionQueue.Push(item)
+				}
+
+				c.backupMu.Lock()
+				delete(c.summaries, chanID)
+				delete(c.chanCommitHeights, chanID)
+				c.backupMu.Unlock()
+
+			default:
+			}
+
+		case <-c.forceQuit:
+			return
+
+		case <-c.quit:
+			return
+		}
+	}
+}
+
+// handleClosableSessions listens for new block notifications. For each block,
+// it checks the closableSessionQueue to see if there is a closable session with
+// a delete-height smaller than or equal to the new block, if there is then the
+// tower is informed that it can delete the session and then we also delete it
+// from our DB.
+func (c *TowerClient) handleClosableSessions(
+	blocksChan *chainntnfs.BlockEpochEvent) {
+
+	defer c.wg.Done()
+
+	c.log.Tracef("Starting closable sessions handler")
+	defer c.log.Tracef("Stopping closable sessions handler")
+
+	for {
+		select {
+		case newBlock := <-blocksChan.Epochs:
+			if newBlock == nil {
+				return
+			}
+
+			height := uint32(newBlock.Height)
+			for {
+				select {
+				case <-c.quit:
+					return
+				default:
+				}
+
+				item := c.closableSessionQueue.Top()
+				if item == nil {
+					break
+				}
+
+				if item.deleteHeight > height {
+					break
+				}
+
+				c.closableSessionQueue.Pop()
+
+				// Fetch the session from the DB so that we can
+				// extract the Tower info.
+				sess, err := c.cfg.DB.GetClientSession(
+					item.sessionID,
+				)
+				if err != nil {
+					log.Error("error calling "+
+						"GetClientSession: %v", err)
+					continue
+				}
+
+				err = c.deleteSessionFromTower(sess)
```

Comment:
I guess this can happen if the tower is offline? If so, would the deletion be tried again? I think this would only happen once lnd (WT client) is restarted, which doesn't seem too bad.